### PR TITLE
Phatgoose dev

### PR DESCRIPTION
**Pull request by leopoldwhite**
adding a new method, [Learning to Route Among Specialized Experts for Zero-Shot Generalization](https://arxiv.org/abs/2402.05859) (PHATGOOSE)

adding the corresponding section in user_readme

### CODE REVIEW

**Comment by BunsenFeng** [comment]
trying to run, for this: - **For `train_expert_gate` mode only:** - `expert_path`, no default (required): path to the single LoRA expert directory.

Do you mean the lora adapters have to be saved in a local directory? If it is passed in like this ("model_names": [
    "bunsenfeng/yuru_qw_wizardlm",
    "bunsenfeng/yuru_qw_sharegpt",
    "bunsenfeng/yuru_qw_oasst1"
]) like other weight-based methods, could you save the models to a local directory (something in logs/...) in your code so people don't need to do on their end?

Also, currently the too many modes are a bit daunting for users... maybe strongly recommend one mode in the readme?

**Comment by leopoldwhite** [comment]
> trying to run, for this: - **For `train_expert_gate` mode only:** - `expert_path`, no default (required): path to the single LoRA expert directory.
> 
> Do you mean the lora adapters have to be saved in a local directory? If it is passed in like this ("model_names": [ "bunsenfeng/yuru_qw_wizardlm", "bunsenfeng/yuru_qw_sharegpt", "bunsenfeng/yuru_qw_oasst1" ]) like other weight-based methods, could you save the models to a local directory (something in logs/...) in your code so people don't need to do on their end?
> 
> Also, currently the too many modes are a bit daunting for users... maybe strongly recommend one mode in the readme?

I have updated the phatgoose_dev branch:

- Now we can pass model names directly like other weight-based methods (e.g., `"model_names": ["bunsenfeng/yuru_qw_wizardlm", "bunsenfeng/yuru_qw_sharegpt"]`). The code automatically downloads adapters from HuggingFace Hub to `logs/adapter_cache/`. It can also load local LoRA checkpoints.

- I removed some features originally used for development and debugging (including the `train_expert_gate` mode) and modified how some hyperparameters are set. Corresponding changes have also been made in the readme.

- The current code can run without errors (Use `mode: train_and_infer`), and the loss for training the gate also decreases as we increasing the gate_steps. However, I've found that the final performance is not good. I am still checking whether there is an issue with the implementation or if the method itself makes it difficult to find a suitable set of hyperparameters.

**Comment by BunsenFeng** [comment]
sounds good, testing with train_and_infer

**Review by BunsenFeng — APPROVED** [review]
all done and tested, well done! this was not easy, 辛苦了
